src/librustc/traits/query/type_op/mod.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use infer::canonical::{
+    Canonical, Canonicalized, CanonicalizedQueryResult, QueryRegionConstraint, QueryResult,
+};
+use infer::{InferCtxt, InferOk};
+use std::fmt;
+use std::rc::Rc;
+use traits::query::Fallible;
+use traits::ObligationCause;
+use ty::fold::TypeFoldable;
+use ty::{Lift, ParamEnv, TyCtxt};
+
+pub mod custom;
+pub mod eq;
+pub mod normalize;
+pub mod outlives;
+pub mod prove_predicate;
+pub mod subtype;
+
+pub trait TypeOp<'gcx, 'tcx>: Sized + fmt::Debug {
+    type Output;
+
+    /// Processes the operation and all resulting obligations,
+    /// returning the final result along with any region constraints
+    /// (they will be given over to the NLL region solver).
+    fn fully_perform(
+        self,
+        infcx: &InferCtxt<'_, 'gcx, 'tcx>,
+    ) -> Fallible<(Self::Output, Option<Rc<Vec<QueryRegionConstraint<'tcx>>>>)>;
+}
+
+pub trait QueryTypeOp<'gcx: 'tcx, 'tcx>: fmt::Debug + Sized {
+    type QueryKey: TypeFoldable<'tcx> + Lift<'gcx>;
+    type QueryResult: TypeFoldable<'tcx> + Lift<'gcx>;
+
+    /// Either converts `self` directly into a `QueryResult` (for
+    /// simple cases) or into a `QueryKey` (for more complex cases
+    /// where we actually have work to do).
+    fn prequery(self, tcx: TyCtxt<'_, 'gcx, 'tcx>) -> Result<Self::QueryResult, Self::QueryKey>;
+
+    fn param_env(key: &Self::QueryKey) -> ParamEnv<'tcx>;
+
+    fn perform_query(
+        tcx: TyCtxt<'_, 'gcx, 'tcx>,
+        canonicalized: Canonicalized<'gcx, Self::QueryKey>,
+    ) -> Fallible<CanonicalizedQueryResult<'gcx, Self::QueryResult>>;
+
+    /// "Upcasts" a lifted query result (which is in the gcx lifetime)
+    /// into the tcx lifetime. This is always just an identity cast,
+    /// but the generic code does't realize it, so we have to push the
+    /// operation into the impls that know more specifically what
+    /// `QueryResult` is. This operation would (maybe) be nicer with
+    /// something like HKTs or GATs, since then we could make
+    /// `QueryResult` parametric and `'gcx` and `'tcx` etc.
+    fn upcast_result(
+        lifted_query_result: &'a CanonicalizedQueryResult<'gcx, Self::QueryResult>,
+    ) -> &'a Canonical<'tcx, QueryResult<'tcx, Self::QueryResult>>;
```

Comment:
I would avoid the word `upcast` around lifetimes - it's the opposite of `lift` due to how subtyping works on lifetimes, but the English word suggests the same (positive vertical) direction in both cases.

Also, this seems like a hack - what prevents, at all, a default body for this method? The comment above it hints to there being a reason but it's not clear what that reason might be. How much of that is hiding in type aliases that contain associated type projections?

Reply:
> I would avoid the word upcast around lifetimes - it's the opposite of lift due to how subtyping works on lifetimes, but the English word suggests the same (positive vertical) direction in both cases.

Hmm, good point. What should I call it do you think? I guess maybe "convert_to_tcx" or something.

> Also, this seems like a hack - what prevents, at all, a default body for this method? 

One person's hack is another man's brilliant solution. =) What prevents this from working on the trait case is that we do not know that `<Self::QueryResult as Lift<'gcx>>::Lifted` is a subtype of `Self::QueryResult` (but in fact it always is). So -- in the generic code -- we have a `Lifted<'gcx, Self::QueryResult>` and we need to convert it to `Self::QueryResult`, and hence we invoke this method. I'll improve the comment.